### PR TITLE
Implemented GAT using a sparse adjacency matrix

### PR DESCRIPTION
**Pull request by adocherty**
GAT has been implemented using sparse matrix operations. This only works with a tensorflow backend.

Now, GCN and GAT can both use either dense or sparse adjacency matrices.  The dense adjacency matrices are currently required for the saliency map calculations.

### Review Checklist:
- [x]  Required functionality implemented
- [x]  Unit tests run
- [x]  Documentation matches implementation (doc strings, README, Sphinx)
- [x]  Python notebooks in demos directory run (GAT & Ensemble demos are affected)

### CODE REVIEW

File: demos/node-classification/gat/gat-cora-example.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def train(
     target_name="subject",
 ):
     """
-    Train a GAT model on the specified graph G with given parameters, evaluate it, and save the model.
+    Train a GAT model on the specified graph G with given parameters, zevaluate it, and save the model.
```
Comment (PantelisElinas) [inline demos/node-classification/gat/gat-cora-example.py]: Spelling error, "zevaluate" --> "evaluate"

File: stellargraph/mapper/node_mappers.py [hunk 2]
```diff
@@ -607,24 +714,27 @@ def __init__(self, G, name=None, method="gcn", k=1, sparse=True, transform=None)
...
+        self.use_sparse = sparse
 
-        if not self.sparse:
-            self.Aadj = self.Aadj.todense()
+        # If sparse is specified, check that the backend is tensorflow
+        if self.use_sparse and K.backend() != "tensorflow":
+            warnings.warn(
+                "Sparse adjacency matrices are only supported in tensorflow."
+                " Falling back to using a dense adjacency matrix."
+            )
```
Comment (PantelisElinas) [inline stellargraph/mapper/node_mappers.py]: Should self.use_sparse be set to False after the warning?

File: stellargraph/layer/graph_attention.py [hunk 3]
```diff
@@ -208,38 +232,37 @@ def build(self, input_shape):
...
+        Notes:
+            This does not add self loops to the adjacency matrix.
+            The output indices are only used when `final_layer=True`
 
+        Args:
+            inputs (list): list of inputs with 3 items:
+            node features (size b x N x F),
+            output indices (size b x M),
+            graph adjacency matrix (size b x N x N),
+            where N is the number of nodes in the graph,
```
Comment (sktzwhj) [inline stellargraph/layer/graph_attention.py]: It might be clearer to say what b represents here.

File: stellargraph/layer/graph_attention.py [hunk 4]
```diff
@@ -297,36 +320,217 @@ def call(self, inputs, **kwargs):
...
+            kernel = self.kernels[head]  # W in the paper (F x F')
+            attention_kernel = self.attn_kernels[
+                head
+            ]  # Attention kernel a in the paper (2F' x 1)
+
+            # Compute inputs to attention network
+            features = K.dot(X, kernel)  # (N x F')
+
+            # Compute feature combinations
+            # Note: [[a_1], [a_2]]^T [[Wh_i], [Wh_2]] = [a_1]^T [Wh_i] + [a_2]^T [Wh_j]
```
Comment (sktzwhj) [inline stellargraph/layer/graph_attention.py]: Wh_2 -> Wh_j?

File: stellargraph/mapper/node_mappers.py (hunk 2 above)
Comment (adocherty) [reply stellargraph/mapper/node_mappers.py]: Good catch .. yes we should set the flag to False.

**Comment by adocherty** [comment]
> The changes look good to me. In terms of the changes we may have for the ongoing work, my understanding is that we will still only support the dense version for saliency maps due to the potential edge constraint. Is that correct?

Yes, we are keeping the dense versions of both GCN and GAT which we will use for saliency maps.

File: stellargraph/layer/graph_attention.py (hunk 3 above)
Comment (adocherty) [reply stellargraph/layer/graph_attention.py]: Good point, I'll add a comment.

File: stellargraph/layer/graph_attention.py (hunk 4 above)
Comment (adocherty) [reply stellargraph/layer/graph_attention.py]: Well spotted!

File: demos/node-classification/gat/gat-cora-example.py (hunk 1 above)
Comment (adocherty) [reply demos/node-classification/gat/gat-cora-example.py]: Fixed.

File: stellargraph/mapper/node_mappers.py (hunk 2 above)
Comment (adocherty) [reply stellargraph/mapper/node_mappers.py]: I will note it's hard to test this in a unit test framework as we would have to change the Keras backend for that test.

**Pull request merged**